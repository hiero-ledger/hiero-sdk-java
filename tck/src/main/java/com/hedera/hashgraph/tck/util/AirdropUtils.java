// SPDX-License-Identifier: Apache-2.0
package com.hedera.hashgraph.tck.util;

import com.hedera.hashgraph.sdk.AccountId;
import com.hedera.hashgraph.sdk.NftId;
import com.hedera.hashgraph.sdk.TokenAirdropTransaction;
import com.hedera.hashgraph.sdk.TokenId;
import com.hedera.hashgraph.tck.exception.InvalidJSONRPC2ParamsException;
import com.hedera.hashgraph.tck.methods.sdk.param.transfer.TransferParams;

// Utility class for handling airdrop parameters
public class AirdropUtils {

    /**
     * Handles a single airdrop transfer parameter and adds it to the transaction
     *
     * @param transaction The token airdrop transaction
     * @param transferParam The transfer parameter to handle
     * @throws Exception If an error occurs during processing
     */
    public static void handleAirdropParam(TokenAirdropTransaction transaction, TransferParams transferParam)
            throws Exception {
        if (transferParam.getToken().isPresent()) {
            handleAirdropTokenTransfer(transaction, transferParam);
        } else if (transferParam.getNft().isPresent()) {
            handleAirdropNftTransfer(transaction, transferParam);
        } else {
            throw new InvalidJSONRPC2ParamsException("Invalid transfer parameter");
        }
    }

    /**
     * Handles a token transfer parameter
     *
     * @param transaction The token airdrop transaction
     * @param transferParam The transfer parameter containing token details
     * @throws Exception If an error occurs during processing
     */
    private static void handleAirdropTokenTransfer(TokenAirdropTransaction transaction, TransferParams transferParam)
            throws Exception {
        var token = transferParam.getToken().get();

        AccountId accountId = AccountId.fromString(
                token.getAccountId().orElseThrow(() -> new InvalidJSONRPC2ParamsException("AccountId is required")));

        TokenId tokenId = TokenId.fromString(
                token.getTokenId().orElseThrow(() -> new InvalidJSONRPC2ParamsException("TokenId is required")));

        long amount;
        try {
            amount = Long.parseLong(
                    token.getAmount().orElseThrow(() -> new InvalidJSONRPC2ParamsException("Amount is required")));
        } catch (NumberFormatException e) {
            throw new InvalidJSONRPC2ParamsException("Invalid amount format");
        }

        boolean isApproved = transferParam.getApproved().isPresent()
                && transferParam.getApproved().get();

        if (token.getDecimals().isPresent()) {
            int decimals = token.getDecimals().get().intValue();
            if (isApproved) {
                transaction.addApprovedTokenTransferWithDecimals(tokenId, accountId, amount, decimals);
            } else {
                transaction.addTokenTransferWithDecimals(tokenId, accountId, amount, decimals);
            }
        } else {
            if (isApproved) {
                transaction.addApprovedTokenTransfer(tokenId, accountId, amount);
            } else {
                transaction.addTokenTransfer(tokenId, accountId, amount);
            }
        }
    }

    /**
     * Handles an NFT transfer parameter
     *
     * @param transaction The token airdrop transaction
     * @param transferParam The transfer parameter containing NFT details
     * @throws Exception If an error occurs during processing
     */
    private static void handleAirdropNftTransfer(TokenAirdropTransaction transaction, TransferParams transferParam)
            throws Exception {
        var nft = transferParam.getNft().get();

        AccountId senderAccountId = AccountId.fromString(nft.getSenderAccountId()
                .orElseThrow(() -> new InvalidJSONRPC2ParamsException("SenderAccountId is required")));

        AccountId receiverAccountId = AccountId.fromString(nft.getReceiverAccountId()
                .orElseThrow(() -> new InvalidJSONRPC2ParamsException("ReceiverAccountId is required")));

        long serialNumber;
        try {
            serialNumber = Long.parseLong(nft.getSerialNumber()
                    .orElseThrow(() -> new InvalidJSONRPC2ParamsException("SerialNumber is required")));
        } catch (NumberFormatException e) {
            throw new InvalidJSONRPC2ParamsException("Invalid serial number format");
        }

        TokenId tokenId = TokenId.fromString(
                nft.getTokenId().orElseThrow(() -> new InvalidJSONRPC2ParamsException("TokenId is required")));

        NftId nftId = new NftId(tokenId, serialNumber);

        boolean isApproved = transferParam.getApproved().isPresent()
                && transferParam.getApproved().get();

        if (isApproved) {
            transaction.addApprovedNftTransfer(nftId, senderAccountId, receiverAccountId);
        } else {
            transaction.addNftTransfer(nftId, senderAccountId, receiverAccountId);
        }
    }
}
