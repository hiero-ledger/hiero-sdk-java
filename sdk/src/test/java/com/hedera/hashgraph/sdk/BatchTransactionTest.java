// SPDX-License-Identifier: Apache-2.0
package com.hedera.hashgraph.sdk;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

import io.github.jsonSnapshot.SnapshotMatcher;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

class BatchTransactionTest {

    private static final PrivateKey privateKeyED25519 = PrivateKey.fromString(
            "302e020100300506032b657004220420db484b828e64b2d8f12ce3c0a0e93a0b8cce7af1bb8f39c97732394482538e10");

    private static final PrivateKey privateKeyECDSA =
            PrivateKey.fromStringECDSA("7f109a9e3b0d8ecfba9cc23a3614433ce0fa7ddcc80f2a8f10b222179a5a80d6");

    static final Instant validStart = Instant.ofEpochSecond(1554158542);

    private static final List<Transaction> INNER_TRANSACTIONS = List.of(
            spawnTestTransactionAccountCreate(),
            spawnTestTransactionAccountCreate(),
            spawnTestTransactionAccountCreate());

    private static AccountCreateTransaction spawnTestTransactionAccountCreate() {
        return new AccountCreateTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .setKeyWithAlias(privateKeyECDSA)
                .setKeyWithAlias(privateKeyED25519, privateKeyECDSA)
                .setKeyWithoutAlias(privateKeyED25519)
                .setInitialBalance(Hbar.fromTinybars(450))
                .setAccountMemo("some memo")
                .setReceiverSignatureRequired(true)
                .setAutoRenewPeriod(Duration.ofHours(10))
                .setStakedAccountId(AccountId.fromString("0.0.3"))
                .setAlias("0x5c562e90feaf0eebd33ea75d21024f249d451417")
                .setMaxAutomaticTokenAssociations(100)
                .setMaxTransactionFee(Hbar.fromTinybars(100_000))
                .setBatchKey(privateKeyECDSA)
                .freeze()
                .sign(privateKeyED25519);
    }

    private BatchTransaction spawnTestTransaction() {
        var batchKey = PrivateKey.generateECDSA();

        return new BatchTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .setInnerTransactions(INNER_TRANSACTIONS)
                .freeze()
                .sign(batchKey);
    }

    @BeforeAll
    public static void beforeAll() {
        SnapshotMatcher.start(Snapshot::asJsonString);
    }

    @AfterAll
    public static void afterAll() {
        SnapshotMatcher.validateSnapshots();
    }

    @Test
    void shouldSerialize() {
        SnapshotMatcher.expect(spawnTestTransaction().toString()).toMatchSnapshot();
    }

    @Test
    void shouldBytes() throws Exception {
        var tx = spawnTestTransaction();
        var tx2 = BatchTransaction.fromBytes(tx.toBytes());
        assertThat(tx2).hasToString(tx.toString());
    }

    @Test
    void shouldBytesNoSetters() throws Exception {
        var tx = new BatchTransaction();
        var tx2 = BatchTransaction.fromBytes(tx.toBytes());
        assertThat(tx2).hasToString(tx.toString());
    }

    @Test
    void getInnerTransactionsShouldReturnCorrectTransactions() {
        var batchTransaction = spawnTestTransaction();
        assertThat(batchTransaction.getInnerTransactions())
                .isNotNull()
                .hasSize(3)
                .isEqualTo(INNER_TRANSACTIONS);
    }

    @Test
    void setInnerTransactionsShouldUpdateTransactions() {
        var batchTransaction = new BatchTransaction();
        List<Transaction> newInnerTransactions =
                List.of(spawnTestTransactionAccountCreate(), spawnTestTransactionAccountCreate());

        batchTransaction.setInnerTransactions(newInnerTransactions);

        assertThat(batchTransaction.getInnerTransactions())
                .isNotNull()
                .hasSize(2)
                .isEqualTo(newInnerTransactions);
    }

    @Test
    void addInnerTransactionShouldAppendTransaction() {
        var batchTransaction = new BatchTransaction();
        var newTransaction = spawnTestTransactionAccountCreate();

        batchTransaction.addInnerTransaction(newTransaction);

        assertThat(batchTransaction.getInnerTransactions())
                .isNotNull()
                .hasSize(1)
                .contains(newTransaction);
    }

    @Test
    void getInnerTransactionIdsShouldReturnCorrectIds() {
        var batchTransaction = spawnTestTransaction();
        var expectedTransactionId = TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart);

        var transactionIds = batchTransaction.getInnerTransactionIds();

        assertThat(transactionIds).isNotNull().hasSize(3).allSatisfy(id -> assertThat(id)
                .isEqualTo(expectedTransactionId));
    }

    @Test
    void shouldAllowChainedSetters() {
        var batchTransaction = new BatchTransaction()
                .setNodeAccountIds(Collections.singletonList(AccountId.fromString("0.0.5005")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .addInnerTransaction(spawnTestTransactionAccountCreate())
                .freeze();

        assertThat(batchTransaction.getInnerTransactions()).hasSize(1);
        assertThat(batchTransaction.getNodeAccountIds()).hasSize(1);
        assertThat(batchTransaction.getTransactionId()).isNotNull();
    }

    @Test
    void shouldRejectFreezeTransaction() {
        var batchTransaction = new BatchTransaction();
        var freezeTransaction = new FreezeTransaction()
                .setStartTime(Instant.now())
                .setFreezeType(FreezeType.FREEZE_ONLY)
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .freeze();

        assertThatExceptionOfType(IllegalArgumentException.class)
                .isThrownBy(() -> batchTransaction.addInnerTransaction(freezeTransaction))
                .withMessageContaining("FreezeTransaction is not allowed in a batch transaction");
    }

    @Test
    void shouldRejectBatchTransaction() {
        var batchTransaction = new BatchTransaction();
        var innerBatchTransaction = new BatchTransaction()
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .freeze();

        assertThatExceptionOfType(IllegalArgumentException.class)
                .isThrownBy(() -> batchTransaction.addInnerTransaction(innerBatchTransaction))
                .withMessageContaining("BatchTransaction is not allowed in a batch transaction");
    }

    @Test
    void shouldRejectBlacklistedTransactionInList() {
        var batchTransaction = new BatchTransaction();
        var validTransaction = spawnTestTransactionAccountCreate();
        var freezeTransaction = new FreezeTransaction()
                .setStartTime(Instant.now())
                .setFreezeType(FreezeType.FREEZE_ONLY)
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .freeze();

        assertThatExceptionOfType(IllegalArgumentException.class)
                .isThrownBy(() -> batchTransaction.setInnerTransactions(List.of(validTransaction, freezeTransaction)))
                .withMessageContaining("FreezeTransaction is not allowed in a batch transaction");
    }

    @Test
    void shouldRejectNullTransaction() {
        var batchTransaction = new BatchTransaction();

        assertThatExceptionOfType(NullPointerException.class)
                .isThrownBy(() -> batchTransaction.addInnerTransaction(null));
    }

    @Test
    void shouldRejectNullTransactionList() {
        var batchTransaction = new BatchTransaction();

        assertThatExceptionOfType(NullPointerException.class)
                .isThrownBy(() -> batchTransaction.setInnerTransactions(null));
    }

    @Test
    void shouldRejectUnfrozenTransaction() {
        var batchTransaction = new BatchTransaction();
        var unfrozenTransaction = new AccountCreateTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart));

        assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(() -> batchTransaction.addInnerTransaction(unfrozenTransaction))
                .withMessageContaining("Inner transaction should be frozen");
    }

    @Test
    void shouldRejectTransactionAfterFreeze() {
        var batchTransaction = new BatchTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .freeze();

        assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(() -> batchTransaction.addInnerTransaction(spawnTestTransactionAccountCreate()))
                .withMessageContaining("transaction is immutable");
    }

    @Test
    void shouldRejectTransactionListAfterFreeze() {
        var batchTransaction = new BatchTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .freeze();

        assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(() -> batchTransaction.setInnerTransactions(INNER_TRANSACTIONS))
                .withMessageContaining("transaction is immutable");
    }

    @Test
    void shouldAllowEmptyTransactionListBeforeExecution() {
        var batchTransaction = new BatchTransaction();
        batchTransaction.setInnerTransactions(Collections.emptyList());

        assertThat(batchTransaction.getInnerTransactions()).isNotNull().isEmpty();
    }

    @Test
    void shouldPreserveTransactionOrder() {
        var batchTransaction = new BatchTransaction();
        var transaction1 = spawnTestTransactionAccountCreate();
        var transaction2 = spawnTestTransactionAccountCreate();
        var transaction3 = spawnTestTransactionAccountCreate();

        List<Transaction> transactions = Arrays.asList(transaction1, transaction2, transaction3);
        batchTransaction.setInnerTransactions(transactions);

        assertThat(batchTransaction.getInnerTransactions()).containsExactly(transaction1, transaction2, transaction3);
    }

    @Test
    void shouldCreateDefensiveCopyOfTransactionList() {
        var batchTransaction = new BatchTransaction();
        var mutableList = new ArrayList<>(INNER_TRANSACTIONS);

        batchTransaction.setInnerTransactions(mutableList);
        mutableList.clear();

        assertThat(batchTransaction.getInnerTransactions())
                .isNotNull()
                .hasSize(3)
                .isEqualTo(INNER_TRANSACTIONS);
    }

    @Test
    void shouldRejectTransactionWithoutBatchKey() {
        var batchTransaction = new BatchTransaction();
        var transactionWithoutBatchKey = new AccountCreateTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .freeze();

        assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(() -> batchTransaction.addInnerTransaction(transactionWithoutBatchKey))
                .withMessageContaining("Batch key needs to be set");
    }

    @Test
    void shouldValidateAllTransactionsInList() {
        var batchTransaction = new BatchTransaction();
        var validTransaction = spawnTestTransactionAccountCreate();
        var transactionWithoutBatchKey = new AccountCreateTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .freeze();

        assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(() ->
                        batchTransaction.setInnerTransactions(List.of(validTransaction, transactionWithoutBatchKey)))
                .withMessageContaining("Batch key needs to be set");
    }

    @Test
    void shouldValidateMultipleConditions() {
        var batchTransaction = new BatchTransaction();

        // Test unfrozen transaction with no batch key
        var unfrozenTransactionWithoutBatchKey = new AccountCreateTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart));

        assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(() -> batchTransaction.addInnerTransaction(unfrozenTransactionWithoutBatchKey))
                .withMessageContaining("Inner transaction should be frozen");

        // Test frozen transaction with no batch key
        var frozenTransactionWithoutBatchKey = unfrozenTransactionWithoutBatchKey.freeze();

        assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(() -> batchTransaction.addInnerTransaction(frozenTransactionWithoutBatchKey))
                .withMessageContaining("Batch key needs to be set");

        // Test blacklisted transaction with batch key
        var blacklistedTransaction = new FreezeTransaction()
                .setStartTime(Instant.now())
                .setFreezeType(FreezeType.FREEZE_ONLY)
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .setBatchKey(privateKeyECDSA)
                .freeze();

        assertThatExceptionOfType(IllegalArgumentException.class)
                .isThrownBy(() -> batchTransaction.addInnerTransaction(blacklistedTransaction))
                .withMessageContaining("FreezeTransaction is not allowed in a batch transaction");
    }

    @Test
    void shouldAcceptValidTransaction() {
        var batchTransaction = new BatchTransaction();
        var validTransaction = new AccountCreateTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart))
                .setBatchKey(privateKeyECDSA)
                .freeze();

        batchTransaction.addInnerTransaction(validTransaction);

        assertThat(batchTransaction.getInnerTransactions())
                .isNotNull()
                .hasSize(1)
                .contains(validTransaction);
    }

    @Test
    void shouldValidateTransactionStateInOrder() {
        var batchTransaction = new BatchTransaction();
        var transaction = new AccountCreateTransaction()
                .setNodeAccountIds(Arrays.asList(AccountId.fromString("0.0.5005"), AccountId.fromString("0.0.5006")))
                .setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), validStart));

        // First check should be for frozen state
        assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(() -> batchTransaction.addInnerTransaction(transaction))
                .withMessageContaining("Inner transaction should be frozen");

        // After freezing, next check should be for batch key
        var frozenTransaction = transaction.freeze();
        assertThatExceptionOfType(IllegalStateException.class)
                .isThrownBy(() -> batchTransaction.addInnerTransaction(frozenTransaction))
                .withMessageContaining("Batch key needs to be set");
    }
}
