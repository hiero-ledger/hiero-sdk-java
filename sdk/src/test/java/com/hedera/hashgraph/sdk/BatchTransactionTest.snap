com.hedera.hashgraph.sdk.BatchTransactionTest.shouldSerialize=[
  "# com.hedera.hashgraph.sdk.proto.TransactionBody\natomic_batch {\n  transactions: \"\\n\\233\\001\\n\\r\\n\\006\\b\\316\\247\\212\\345\\005\\022\\003\\030\\216\\'\\022\\003\\030\\215\\'\\030\\240\\215\\006\\\"\\002\\bxZW\\n\\\"\\022 \\340\\310\\354\\'X\\245\\207\\237\\372\\302&\\241<\\fQky\\236r\\343QA\\240\\335\\202\\217\\224\\323y\\210\\244\\267\\020\\302\\003@\\001J\\004\\b\\240\\231\\002j\\tsome memopdz\\002\\030\\003\\222\\001\\024\\\\V.\\220\\376\\257\\016\\353\\323>\\247]!\\002O$\\235E\\024\\027\\312\\004#:!\\002Gx\\024\\037\\377\\344b\\333\\263\\201|\\223\\344%+-o_Vg\\372q\\363K\\204(\\002e\\300\\035@\\217\\022f\\nd\\n \\340\\310\\354\\'X\\245\\207\\237\\372\\302&\\241<\\fQky\\236r\\343QA\\240\\335\\202\\217\\224\\323y\\210\\244\\267\\032\\037u\\366\\305\\vr7\\373\\367\\b\\354\\003\\215\\032\\016\\344\\352\\216S%\\256\\\\\\210~0n\\344jJ\\203s\\324\\2569\\310Sk\\354J\\347\\000I\\300\\a\\255\\210P\\235d\\357\\244(F\\324T\\215\\352j \\343\\025\\223\\016\"\n  transactions: \"\\n\\233\\001\\n\\r\\n\\006\\b\\316\\247\\212\\345\\005\\022\\003\\030\\216\\'\\022\\003\\030\\215\\'\\030\\240\\215\\006\\\"\\002\\bxZW\\n\\\"\\022 \\340\\310\\354\\'X\\245\\207\\237\\372\\302&\\241<\\fQky\\236r\\343QA\\240\\335\\202\\217\\224\\323y\\210\\244\\267\\020\\302\\003@\\001J\\004\\b\\240\\231\\002j\\tsome memopdz\\002\\030\\003\\222\\001\\024\\\\V.\\220\\376\\257\\016\\353\\323>\\247]!\\002O$\\235E\\024\\027\\312\\004#:!\\002Gx\\024\\037\\377\\344b\\333\\263\\201|\\223\\344%+-o_Vg\\372q\\363K\\204(\\002e\\300\\035@\\217\\022f\\nd\\n \\340\\310\\354\\'X\\245\\207\\237\\372\\302&\\241<\\fQky\\236r\\343QA\\240\\335\\202\\217\\224\\323y\\210\\244\\267\\032\\037u\\366\\305\\vr7\\373\\367\\b\\354\\003\\215\\032\\016\\344\\352\\216S%\\256\\\\\\210~0n\\344jJ\\203s\\324\\2569\\310Sk\\354J\\347\\000I\\300\\a\\255\\210P\\235d\\357\\244(F\\324T\\215\\352j \\343\\025\\223\\016\"\n  transactions: \"\\n\\233\\001\\n\\r\\n\\006\\b\\316\\247\\212\\345\\005\\022\\003\\030\\216\\'\\022\\003\\030\\215\\'\\030\\240\\215\\006\\\"\\002\\bxZW\\n\\\"\\022 \\340\\310\\354\\'X\\245\\207\\237\\372\\302&\\241<\\fQky\\236r\\343QA\\240\\335\\202\\217\\224\\323y\\210\\244\\267\\020\\302\\003@\\001J\\004\\b\\240\\231\\002j\\tsome memopdz\\002\\030\\003\\222\\001\\024\\\\V.\\220\\376\\257\\016\\353\\323>\\247]!\\002O$\\235E\\024\\027\\312\\004#:!\\002Gx\\024\\037\\377\\344b\\333\\263\\201|\\223\\344%+-o_Vg\\372q\\363K\\204(\\002e\\300\\035@\\217\\022f\\nd\\n \\340\\310\\354\\'X\\245\\207\\237\\372\\302&\\241<\\fQky\\236r\\343QA\\240\\335\\202\\217\\224\\323y\\210\\244\\267\\032\\037u\\366\\305\\vr7\\373\\367\\b\\354\\003\\215\\032\\016\\344\\352\\216S%\\256\\\\\\210~0n\\344jJ\\203s\\324\\2569\\310Sk\\354J\\347\\000I\\300\\a\\255\\210P\\235d\\357\\244(F\\324T\\215\\352j \\343\\025\\223\\016\"\n}\nnode_account_i_d {\n  account_num: 5005\n  realm_num: 0\n  shard_num: 0\n}\ntransaction_fee: 200000000\ntransaction_i_d {\n  account_i_d {\n    account_num: 5006\n    realm_num: 0\n    shard_num: 0\n  }\n  transaction_valid_start {\n    seconds: 1554158542\n  }\n}\ntransaction_valid_duration {\n  seconds: 120\n}"
]